### PR TITLE
feat(pics): remove variants from `locus` when PICS cannot be applied

### PR DESCRIPTION
**Pull request by ireneisdoomed**
This PR contains:
- Changes in the PICS logic that removes variants in the ldSet that don't fit the criteria to calculate PIPs with PICS. I initially decided to keep these variants before the concept of `ldSet` and `locus` were split. Now that we have both structs, we don't need to maintain the steadiness within the locus.
- Addition of a QC flag when "locus" is empty
- Minor change: addition of `pics` in fineMappingMethod

## Context
While doing L2G stuff, I saw this association. None of the variants in the ldSet after weighting surpass the 0.5 threshold to run PICS on.
```
------------------------------------------------------------------------------------------------------------
 studyId           | GCST008046                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
 variantId         | 2_74288907_T_C                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
 ldSet             | [{2_74596587_GAAC_G, 0.1663480955127811}, {2_74385832_T_C, 0.2955881348428679}, {2_74501008_G_C, 0.2893655552523453}, {2_74158608_G_A, 0.1845705771149719}, {2_74726236_C_T, 0.2378922729310847}, {2_74689824_A_G, 0.2379001711729858}, {2_74731163_T_C, 0.23790017117298576}, {2_74329398_T_C, 0.28328273017567607}, {2_74288907_T_C, 0.3015561263030666}, {2_74210611_A_G, 0.19195220062794693}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
 locus             | [{true, true, null, null, 2_74596587_GAAC_G, null, null, null, null, null, null, null, null, 0.1663480955127811}, {false, false, null, null, 2_74385832_T_C, null, null, null, null, null, null, null, null, 0.2955881348428679}, {false, false, null, null, 2_74501008_G_C, null, null, null, null, null, null, null, null, 0.2893655552523453}, {false, false, null, null, 2_74158608_G_A, null, null, null, null, null, null, null, null, 0.1845705771149719}, {false, false, null, null, 2_74726236_C_T, null, null, null, null, null, null, null, null, 0.2378922729310847}, {false, false, null, null, 2_74689824_A_G, null, null, null, null, null, null, null, null, 0.2379001711729858}, {false, false, null, null, 2_74731163_T_C, null, null, null, null, null, null, null, null, 0.23790017117298576}, {false, false, null, null, 2_74329398_T_C, null, null, null, null, null, null, null, null, 0.28328273017567607}, {false, false, null, null, 2_74288907_T_C, null, null, null, null, null, null, null, null, 0.3015561263030666}, {false, false, null, null, 2_74210611_A_G, null, null, null, null, null, null, null, null, 0.19195220062794693}] 
 fineMappingMethod | null                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
 qualityControls   | []       
```

Looking at the PICS UDF function, we have 2 arrays:
- new_credible_set: the variants that are going to shape the new `locus`. 
- tmp_credible_set: where we store the variants that can be PICSed and that will eventually promote to new_credible_set

When the variants were not eligible for PICS, I added them directly to new_credible_set so that I don't lose them. This made sense when the `ldSet` was a dynamic field, but now that it is split with `locus`, I don't think it makes sense keeping variants without PIPs.
Now the same association looks like this:
```
-RECORD 0---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 studyId           | GCST008046                                                                                                                                                                                                                                                                                                                                                                                         
 variantId         | 2_74288907_T_C                                                                                                                                                                                                                                                                                                                                                                                     
 ldSet             | [{2_74596587_GAAC_G, 0.1663480955127811}, {2_74385832_T_C, 0.2955881348428679}, {2_74501008_G_C, 0.2893655552523453}, {2_74158608_G_A, 0.1845705771149719}, {2_74726236_C_T, 0.2378922729310847}, {2_74689824_A_G, 0.2379001711729858}, {2_74731163_T_C, 0.23790017117298576}, {2_74329398_T_C, 0.28328273017567607}, {2_74288907_T_C, 0.3015561263030666}, {2_74210611_A_G, 0.19195220062794693}] 
 locus             | null                                                                                                                                                                                                                                                                                                                                                                                               
 fineMappingMethod | pics                                                                                                                                                                                                                                                                                                                                                                                               
 qualityControls   | [Unable to calculate PIPs with the provided data] 
```

### CODE REVIEW

File: src/otg/method/pics.py [hunk 1]
```diff
@@ -146,7 +146,6 @@ def _finemap(
                 or not lead_neglog_p
             ):
                 # If PICS cannot be calculated, we'll return the original credible set
```
Comment (d0choa) [inline src/otg/method/pics.py:148]: to be removed

File: src/otg/method/pics.py [hunk 2]
```diff
@@ -239,7 +238,18 @@ def finemap(
                         ),
                     ),
                 )
-                # Rename tagVariantId to variantId
+                .withColumn(
+                    "qualityControls",
+                    StudyLocus.update_quality_flag(
+                        f.col("qualityControls"),
+                        f.size("locus") == 0,
```
Comment (d0choa) [inline src/otg/method/pics.py:245]: can we filter ldset and check valid R-squared variants?

File: src/otg/dataset/study_locus.py [hunk 3]
```diff
@@ -49,6 +50,7 @@ class StudyLocusQualityCheck(Enum):
     UNRESOLVED_LD = "Variant not found in LD reference"
     LD_CLUMPED = "Explained by a more significant variant in high LD (clumped)"
     NO_POPULATION = "Study does not have population annotation to resolve LD"
+    UNPICSABLE = "Unable to calculate PIPs with the provided data"
```
Comment (d0choa) [inline src/otg/dataset/study_locus.py:53]:
```suggestion
    NOT_QUALIFYING_LD_BLOCK = "LD block does not contain variants at the required R^2 threshold"
```

File: src/otg/method/pics.py [hunk 4]
```diff
@@ -239,7 +241,18 @@ def finemap(
                         ),
                     ),
                 )
-                # Rename tagVariantId to variantId
+                .withColumn(
+                    "qualityControls",
+                    StudyLocus.update_quality_flag(
+                        f.col("qualityControls"),
+                        non_picsable_expr,
+                        StudyLocusQualityCheck.UNPICSABLE,
```
Comment (d0choa) [inline src/otg/method/pics.py:249]:
```suggestion
                        StudyLocusQualityCheck. NOT_QUALIFYING_LD_BLOCK,
```

File: tests/method/test_pics.py [hunk 5]
```diff
@@ -30,11 +30,26 @@ def test_finemap_empty_array(
...
+        self: TestFinemap, mock_study_locus: StudyLocus
+    ) -> None:
+        """Test that we add a `empty locus` flag when any variant in the locus meets PICS criteria."""
+        mock_study_locus.df = mock_study_locus.df.withColumn(
+            # Association with an empty ldSet
+            "ldSet",
+            f.when(f.col("ldSet").isNull(), f.array()).otherwise(f.col("ldSet")),
+        ).filter(f.size("ldSet") == 0)
+        observed_df = PICS.finemap(mock_study_locus).df.limit(1)
+        qc_flag = "Unable to calculate PIPs with the provided data"
```
Comment (d0choa) [inline tests/method/test_pics.py:48]:
```suggestion
        qc_flag = "LD block does not contain variants at the required R^2 threshold"
```

**Review by d0choa — APPROVED** [review]

**Review by DSuveges — APPROVED** [review]

**Pull request merged**